src/Raven.Server/Rachis/CandidateAmbassador.cs

```diff
@@ -109,12 +109,19 @@ private void Run()
                         Action disconnect;
                         try
                         {
-                            var connection = _engine.ConnectToPeer(_url, _tag, _engine.ClusterCertificate).Result;
+                            var connectTask = _engine.ConnectToPeer(_url, _tag, _engine.ClusterCertificate);
+                            while (connectTask.Wait(1000) == false)
```

Comment:
Why are we doing it in this manner?
And can we use a cancellation token here, instead?

Reply:
the await is persisted from basically three parts:

1. fetch tcp info (can pass a cancellation token to the request executor)
2. tcp client tries to connect (haven't saw an overload with cancellation)
3. wrap ssl stream (no token either).

Reply:
See: https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.wait?view=netframework-4.7.2#System_Threading_Tasks_Task_Wait_System_Int32_System_Threading_CancellationToken_

Reply:
and for (1) we cancel it after 15 seconds.

Reply:
Okay... I thought of passing a token to the `ConnectToPeer` method.
I can pass a token, but I will still need to validate the term (more relevant for the follower ambassador)

Reply:
We will deal with it in a different issue by limiting the concurrent number of ambassadors:
https://issues.hibernatingrhinos.com/issue/RavenDB-13067